wicketd/src/installinator_progress.rs

```diff
@@ -69,34 +78,42 @@ impl IprArtifactServer {
                     );
                     let (sender, receiver) = mpsc::channel(16);
                     _ = start_sender.send(receiver);
-                    *update =
-                        RunningUpdate::send_and_next_state(sender, report)
-                            .await;
+                    *update = RunningUpdate::next_state(&sender, &report);
+                    Some(sender)
                 }
                 RunningUpdate::ReportsReceived(sender) => {
                     slog::debug!(
                         self.log,
                         "further report seen for this update ID";
                         "update_id" => %update_id
                     );
-                    *update =
-                        RunningUpdate::send_and_next_state(sender, report)
-                            .await;
+                    *update = RunningUpdate::next_state(&sender, &report);
+                    Some(sender)
                 }
                 RunningUpdate::Closed => {
                     // The sender has been closed; ignore the report.
                     *update = RunningUpdate::Closed;
+                    None
                 }
                 RunningUpdate::Invalid => {
-                    unreachable!("invalid state")
+                    unreachable!("invalid state");
                 }
-            }
+            };
 
-            EventReportStatus::Processed
+            (EventReportStatus::Processed, sender)
         } else {
             slog::debug!(self.log, "update ID unrecognized"; "update_id" => %update_id);
-            EventReportStatus::UnrecognizedUpdateId
+            (EventReportStatus::UnrecognizedUpdateId, None)
+        };
+
+        // Don't drop the mutex here because we don't want messages to be sent
+        // out-of-order. (If we were OK with messages being sent out-of-order,
+        // this could be a std Mutex rather than a Tokio mutex.)
+        if let Some(sender) = sender {
+            _ = sender.send(report).await;
```

Comment:
If we are cancelled here, we have lost `report` but have already updated our `*update` to the next state (possibly `Self::Closed`, if this was a terminal report). Is that okay?

Reply:
Good catch, fixed by spawning a task to send the report.

Reply:
Changed this so that it uses an unbounded receiver: that simplifies things substantially.